ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,1439 @@
+  disabled = false,
+}: KeyComboboxProps) {
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  return (
+    <div className="relative w-full">
+      <Combobox immediate onChange={onSelect} disabled={disabled}>
+        {() => (
+          <>
+            <div className="relative">
+              <ComboboxInput
+                ref={inputRef}
+                className={`macro-input ${disabled ? 'opacity-60 cursor-not-allowed' : ''}`}
+                placeholder={disabled ? "Max keys reached" : "Search for key..."}
+                displayValue={() => query}
+                onChange={(event) => onQueryChange(event.target.value)}
+                disabled={disabled}
+              />
+            </div>
+            
+            <ComboboxOptions className="absolute left-0 z-50 mt-1 w-full max-h-60 overflow-auto rounded-md bg-white dark:bg-slate-800 py-1 text-sm shadow-lg">
+              {getFilteredOptions().map((option) => (
+                <ComboboxOption 
+                  key={option.value} 
+                  value={option}
+                  className="cursor-default select-none py-1.5 px-3 ui-active:bg-blue-100 ui-active:text-blue-900 dark:text-slate-300 dark:ui-active:bg-blue-900/40 dark:ui-active:text-blue-200"
+                >
+                  {option.label}
+                </ComboboxOption>
+              ))}
+              {getFilteredOptions().length === 0 && (
+                <div className="py-2 px-3 text-sm text-slate-500 dark:text-slate-400">
+                  No matching keys found
+                </div>
+              )}
+            </ComboboxOptions>
+          </>
+        )}
+      </Combobox>
+    </div>
+  );
+}
+
+const PRESET_DELAYS = [
+  { value: 50, label: "50ms" },
+  { value: 100, label: "100ms" },
+  { value: 200, label: "200ms" },
+  { value: 300, label: "300ms" },
+  { value: 500, label: "500ms" },
+  { value: 750, label: "750ms" },
+  { value: 1000, label: "1000ms" },
+  { value: 1500, label: "1500ms" },
+  { value: 2000, label: "2000ms" },
+];
+
+const MAX_STEPS_PER_MACRO = 10;
+const MAX_TOTAL_MACROS = 25;
+const MAX_KEYS_PER_STEP = 10;
+
+const ensureArray = <T,>(arr: T[] | null | undefined): T[] => {
+  return Array.isArray(arr) ? arr : [];
+};
+
+// Helper function to normalize sort orders, ensuring they start at 1 and have no gaps
+const normalizeSortOrders = (macros: KeySequence[]): KeySequence[] => {
+  return macros.map((macro, index) => ({
+    ...macro,
+    sortOrder: index + 1,
+  }));
+};
+
+interface MacroStepCardProps {
+  step: MacroStep;
+  stepIndex: number;
+  onDelete?: () => void;
+  onMoveUp?: () => void;
+  onMoveDown?: () => void;
+  isDesktop: boolean;
+  onKeySelect: (option: KeyOptionData) => void;
+  onKeyQueryChange: (query: string) => void;
+  keyQuery: string;
+  getFilteredKeys: () => KeyOption[];
+  onModifierChange: (modifiers: string[]) => void;
+  onDelayChange: (delay: number) => void;
+  isLastStep: boolean;
+}
+
+function MacroStepCard({
```

Comment:
Extract into file. We need to cut the size of this file to increase readability

Reply:
extracted into MacroForm and MacroStepCard components